ocaml/jane/doc/extensions/unboxed-types/index.md

```diff
@@ -284,4 +284,144 @@ Here's the list of primitives that currently support `[@layout_poly]`:
+solely of `float` and `float#` fields. Any ordering of `float` and `float#`
+fields is permitted. The "flat float record optimization" applies to any
+such record&mdash;all of the fields are stored flat, even the `float` ones
+that will require boxing upon projection. The ordering restriction is relaxed
+in this case to provide a better migration story for all-`float` records
+to which the flat float record optimization currently applies.
+
+```ocaml
+type t_flat_float =
+  { x1 : float;
+    x2 : float#;
+  }
+```
+
+The ordering restriction has to do with the "mixed block" runtime
+representation. Read on for more detail about that.
+
+## Generic operations aren't supported
+
+Some operations built in to the OCaml runtime aren't supported for structures
+containing unboxed types.
+
+These operations aren't supported:
+  * polymorphic comparison and equality
+  * polymorphic hash
+  * marshaling
+  
+These operations raise an exception at runtime, similar to how polymorphic
+comparison raises when called on a function.
+
+You should use ppx-derived versions of these operations instead.
+
+## Runtime representation: mixed blocks
+
+The general principle guiding the restriction on field ordering in the user
+program: The compiler should not change the user's order of fields
+when deciding the runtime representation.
+
+Abiding by this principle this allows you to predictably write C bindings and
+predict hardware cache performance.
+
+A structure containing mixed blocks is represented at runtime as a "mixed
+block". A mixed block always consists of fields the GC can scan followed by
+fields the GC either could or must skip. The garbage collector must be kept
+informed of which fields of the block it should scan. A portion of the header
+word is reserved to track the length of the prefix of the block that should be
+scanned by the garbage collector.
+
+The ordering constraint on structure fields is a reflection of the same
+ordering restriction in the runtime representation. 
+
+## C bindings for mixed blocks
+
+The implementation of field layout in a mixed block is not finalized. For example, we'd like for int32 fields to be packed efficiently (two to a word) on 64 bit platforms. Currently that's not the case: each one takes up a word.
+
+Users who write C bindings might want to be notified when we change this layout. To ensure that your code will need to be updated when the layout changes, use the `Assert_mixed_block_layout_v#` family of macros. For example,
+
+```
+Assert_mixed_block_layout_v1;
+```
+
+Write the above in statement context&mdash;it should be OK to do so at the top-level of a file or within a function.
+
+Here's a full example. Say you're writing C bindings against this OCaml type:
+
+```ocaml
+(** foo.ml *)
+type t =
+  { x : int32#;
+    y : int32#;
+  }
+```
+
+Here is the recommend way to access fields:
+
+```c
+Assert_mixed_block_layout_v1;
+#define Foo_t_x(foo) (((int32_t*)foo)[0])
+#define Foo_t_y(foo) (((int32_t*)foo)[2])
```

Comment:
Casting it to an int32_t array is a bit weird, would it not be better to do something like `(int32_t*)&Field(foo,0)` and `(int32_t*)&Field(foo,1)`?